File: storage/providers/local/local.go
package local

import (
	"context"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

type LocalStorageProvider struct {
	BaseDir string
	BaseURL string
}

func NewLocalStorageProvider(baseDir, baseURL string) (*LocalStorageProvider, error) {
	if baseDir == "" {
		baseDir = "./uploads"
	}
	if baseURL == "" {
		baseURL = "/uploads"
	}

	if err := os.MkdirAll(baseDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create upload directory '%s': %w", baseDir, err)
	}

	return &LocalStorageProvider{
		BaseDir: baseDir,
		BaseURL: baseURL,
	}, nil
}

func (p *LocalStorageProvider) Upload(ctx context.Context, file io.Reader, filename string, contentType string) (string, error) {
	fullPath := filepath.Join(p.BaseDir, filename)
	dir := filepath.Dir(fullPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", fmt.Errorf("failed to create directory '%s': %w", dir, err)
	}

	out, err := os.Create(fullPath)
	if err != nil {
		return "", fmt.Errorf("failed to create file '%s': %w", fullPath, err)
	}
	defer out.Close()

	if _, err := io.Copy(out, file); err != nil {
		return "", fmt.Errorf("failed to write file content: %w", err)
	}

	fileURL := fmt.Sprintf("%s/%s", p.BaseURL, filepath.ToSlash(filename))
	return fileURL, nil
}

func (p *LocalStorageProvider) Delete(ctx context.Context, fileURL string) error {
	relPath := filepath.Clean(fileURL)
	fullPath := filepath.Join(p.BaseDir, relPath)
	return os.Remove(fullPath)
}

func (p *LocalStorageProvider) GetProviderName() string {
	return "LOCAL"
}

File: backend_api/storage/local.py
import os


class LocalStorageProvider:
    def __init__(self, base_dir: str = "", base_url: str = ""):
        self.base_dir = os.path.abspath(base_dir or "./uploads")
        self.base_url = (base_url or "/uploads").rstrip("/")
        os.makedirs(self.base_dir, exist_ok=True)

    def upload(self, context, file, filename: str, content_type: str) -> str:
        relative = filename.replace("\\", "/").lstrip("/")
        full_path = os.path.join(self.base_dir, relative.replace("/", os.sep))
        os.makedirs(os.path.dirname(full_path), exist_ok=True)

        data = file.read() if hasattr(file, "read") else file
        with open(full_path, "wb") as out:
            out.write(data)

        return f"{self.base_url}/{relative}"

    def delete(self, context, file_url: str):
        url = (file_url or "").replace("\\", "/")
        prefix = self.base_url + "/"
        if url.startswith(prefix):
            relative = url[len(prefix):]
        elif url.startswith("/uploads/"):
            relative = url[len("/uploads/"):]
        else:
            relative = url.lstrip("/")

        full_path = os.path.join(self.base_dir, relative.replace("/", os.sep))
        if os.path.exists(full_path):
            os.remove(full_path)

    def get_provider_name(self) -> str:
        return "LOCAL"
